<a href="https://colab.research.google.com/github/valentinacorreazta-pixel/Anal-tica-de-Negocios-/blob/main/Arboldecision.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**Caso de Estudio**
Una entidad fiannciera (FINTECH) quiere implementar un modelo de árbol de decisión para mejorar la preaprobación de créditos de consumo de sus solicitantes de este tipo de créditos. Para este proceso vamos a utlizar las variables:

* Edad: indica el número de años que posee una perosna, o el tiempo que usted lleva en el sistema financiero.n
* Ingresos: Engloba todos los ingresos que recibe una persona además si posee salario mensual (USD).
* Egresos:
* Monto: (EAD)

0. se procee con la carga de las librerias de trabajo

---



In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

1. se procede con la carga de los datos de trabajo

In [ ]:
nxl='/content/1. SolicitantesCrédito(USD) (2).xlsx'
XDB=pd.read_excel(nxl,sheet_name=1)

XD=XDB.iloc[:,[1,10,11,25]]
yd=XDB.iloc[:,32]
display(XD)
display(yd)

,Edad,Ingresos,Egresos,Monto (EAD)
0,38,1356.14400,1685.622667,625.562230
1,51,286.01600,364.462000,140.031984
2,37,517.46325,629.208889,284.564492
3,29,473.27000,551.616889,309.647588
4,42,750.09175,806.715778,500.663578
...,...,...,...,...
5837,48,1207.84800,753.801111,748.041791
5838,31,1472.77200,953.812889,870.793819
5839,38,773.01975,672.910667,594.947894
5840,43,635.50175,780.691556,305.580539


,PreApr
0,1
1,1
2,0
3,1
4,1
...,...
5837,1
5838,1
5839,1
5840,1


2. Se procede con la implementación del modelo del árbol







In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd)    #Aqui el modelo busca relación entrada-salida

#¿Y que fue lo que hizo el modelo?
ydp=mar.predict(XD)    #Esto es lo que prónostica el modelo

#Se construye la matriz de consfusión
cm=confusion_matrix(yd,ydp)
display(cm)
VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]
#Metricas de Desempeño
Ex=(VP + VN)/len(XD)    #1. Exactitud: Comportamiento General
print("La exactitud es:" ,Ex)
Sen=VP/(VP+FN)     #2. Sensibilidad: Como se comporta pronosticando PreAprobados
print("La sensibilidad es: ",Sen)
Spe=VN/(VN+FP)  #3. Especificidad: Como se comporta frente a los negativos
print("La especificidad")
Pre=VP/(VP+FP)  #4. Precisión: Como se comporta pronosticando PreAprobados
print("La Precición es: ",Pre)
PreNeg=VN/(VN+FN) #5. Precisión Negativa: Como se comporta pronósticando negativos
print("La Predicción Negativa es:",PreNeg)

array([[2301,  658],
       [ 644, 2239]])

La exactitud es: 0.7771311194796303
La sensibilidad es:  0.7766215747485259
La especificidad
La Precición es:  0.772868484639282
La Predicción Negativa es: 0.7813242784380305


3. Despliegue del Árbol de *Decisión*

In [ ]:
from sklearn.tree import export_graphviz     #Exporta los datos a un gráfico
from pydotplus import graph_from_dot_data    #Es un graficador

vs=["Edad", "Ingresos","Egresos", "Monto"]   #Titulos del Árbol
dot_data=export_graphviz(mar,feature_names=vs) #Exportar de números a gráfico
graph=graph_from_dot_data(dot_data)
graph.write_png('Arbol.png')


True

** ANÁLISIS DE RESULTADOS
De la base de datos se puede observar un total de 2883 solicitantes de crédito que poseen la categoría de prenegados, mientras que para la categoría de preaprobados la matriz cuenta con un total de 2883 créditos preaprobados. De los 2959 datos, el modelo pronóstico correctamente un total de 2301 datos. De los 2883 preaprobados el modelos pronostico correctamente un total de 2239. El modelo logra identificar en un 77% los créditos prenegados(2301/2959), mientras que el modelo logró identificar en un 77,66% los creditos aprobados (2239/2883).

Con respecto a las métricas podemos observar que el modelo logró exactitud, 77.71%, lo que indica el buen comportamiento general del modelo frente a la clasificación de créditos en las dos categorías de preaprobación. Se destacan la especifidad y la precisión las cuales lograron valores promedio por encima del 77%, lo que supera el limite inferior definido para este tipo de modelos de clasificación el cual se ubica en el 75%.

De acuerdo con el Árbol de decisión se destaca un nodo puro (10/0) el cual posee la siguiemte regla de decisión, si una persona cumple con esta:
si Monto < = 322 y Ingresos > 376 Y Egre<=685 y Ing >644, tendrá una probabilidad de aprobación del 100%. Es importante destacar que los nodos puros poseen un gini=0 (explicar para el taller los nodos puros y los que estan en los extremos)En general el modelo, diferencia muy bien los buenapagas.

A pesar de la no exixtencia de más nodos puros, se destacan los nodos extremos o las reglas del negocio extremas. se destaca una segunda regla que logra un porcentaje de aprobación del 99%, y la cual posee la siguiente estructura: Si monto <=322 and Ing<=376 and Monto<=178 and Ing<=232 (923/5). ( es decir los nodos son los porcentajes mas altos). Por su parte, el nodo derecho posee un porcentaje de negociación del 98%, en donde la regla se define: si monto>322 and Ing>896 y monto>961 and ingre>1178 (7/576)

Nodo cero= GINI CERO
EXTREMOS= son los nodos puros que dicen 99% ya sea aprobado o rechazado


In [ ]:
from sklearn.tree import export_text

# Nombres de las variables del modelo
vs = ['Pregnancies','Glucose','BloodPressure','SkinThickness',
      'Insulin','BMI','DiabetesPedigreeFunction','Age']

# Mostrar las reglas del árbol
tree_rules = export_text(mar, feature_names=vs)

print("Reglas del árbol de decisión:\n")
print(tree_rules)

NameError: name 'mar' is not defined